### DISCUSSION TITLE
Question: use Fast API Dependency Injection separately from the API routes.

### DISCUSSION THREAD

**Discussion by ievgennaida** [fastapi/fastapi#9108] [Questions]
I have a complicated business logic that dependent on other services (6+ services). 
I would like to test part of the business logic service without running FastAPI TestClient while i need to test only part of it, and not full API call.

Is it possible to reuse build in Dependency Injection  in the Fast API in order to resolve dependent services? 

Example: 
```
class ComputationService:
    def calc():
        # some computations
        return 0
    def transform(i):
        return i+1

class BLService:
    __init__(service:ComputationService = Depends(ComputationService)):
        self.service = service

    def calc_and_transform(self):
        result = self.service.calc()
        transformed = self.service.transform(result)
        return transformed

def unit_test():
       blService = **RESOLVE**(BLService);
       result = blService.calc_and_transform()
       # expected is result 1 
```

Also, is it designed for such use-cases or it's better to use another dependency injection library in parallel?

Thanks in advance.

**Comment by wozniakty** [comment]
I was actually just about to open a similar issue.  It would be pretty incredible if the devs could actually pull the dependency injection concept out of fastAPI and into a standalone library.  I'm genuinely not smart enough to figure out if that's actually possible, but if so it would be a godsend.

An additional reason for this:

Ideally, in support of long-term best-practices, my service logic should be a library that can be wrapped by _any_ API framework (or maybe it just lives in a lambda?  Or a script?  Why should it be coupled to HTTP at all?)

It makes a ton of sense to say that I should essentially have a standalone library for "doing something" and then a FastAPI wrapper around the inputs and outputs of that library to HTTP-ify them.  But doing so today is nearly impossible because I find the `Depends` structure really valuable for dependency resolution, and I use it all the way down the stack for a variety of useful reasons.

**Comment by marianhlavac** [comment]
Have anyone else gave this a thought?

This seems as a serious limitation of FastAPI's dependency injection, in terms of scalability. Passing dependency reference from the route method to all other services kinda defeats the purpose of DI.

**Comment by Arc-xXL** [comment]
I think decoupling the FastAPI DI from the request pipeline would really help in a number of scenarios, however I am not sure if this is even a priority or concern for the guys at FastAPI

I'm gonna post a dirty hack that helped me tap into the DI without having it come from a FastAPI request

```python
from contextlib import AsyncExitStack
from typing import Any
from typing import AsyncGenerator
from typing import Callable
from typing import Coroutine
from typing import Optional

from fastapi.dependencies.utils import get_dependant
from fastapi.dependencies.utils import solve_dependencies
from starlette.requests import Request
from main import app


async def resolve(callable_dep: Callable) -> Any:
    dependent = get_dependant(path="", call=callable_dep)
    values, _, _, _, _ = await solve_dependencies(
        request=__mock_request(),
        dependant=dependent,
        dependency_overrides_provider=app
    )
    resolved = await __create_instance(callable_dep, **values)

    if not resolved:
        raise Exception("failed to create dependency")

    return resolved


def __mock_request() -> Request:
    astack = AsyncExitStack()
    app_instance = {"app": app, "fastapi_astack": astack}
    scope = {"type": "http", "method": "GET", "headers": "", "cookies": "", "query_string": ""}
    return Request({**app_instance, **scope})


async def __create_instance(callable_dep: Callable, **kwargs) -> Optional[Any]:
    instance = callable_dep(**kwargs)
    if isinstance(instance, AsyncGenerator):
        resolved: Optional[Any] = next(iter([i async for i in instance]), None)
    elif isinstance(instance, Coroutine):
        resolved = await instance
    else:
        resolved = instance
    return resolved

```

this is not ideal but it works for anyone who wants to take this and actually do something with it while they figure out a better way to manage their deps

**Reply by ievgennaida** [reply]
I have also endup with the recursive function that is resolving everything, but this one seems to be better while using fast api internal resolver
